ats/pkb/pkb.py

```diff
@@ -44,18 +45,64 @@ def process_relations(node: nodes.ASTNode):
                 if isinstance(parent, nodes.StmtNode):
                     parents[node.__stmt_id] = parent.__stmt_id
 
+            if isinstance(
+                node, (nodes.StmtNode, nodes.StmtLstNode, nodes.ProcedureNode)
+            ):
+                uses_stack.append(node)
+                uses_stack.append([])
+
+            else:
+                uses_stack.append(node)
+
+            if isinstance(node, nodes.VariableNode) and node.parent.variable == node:
+                uses_stack.pop()
+
+            if (
+                isinstance(uses_stack[-1], (nodes.VariableNode, nodes.ConstantNode))
+                and node.parent.left != node
+            ):
+                variable_list = []
+                while not isinstance(uses_stack[-1], list):
+                    if isinstance(uses_stack[-1], nodes.VariableNode):
+                        variable_list.append(uses_stack[-1].name)
+                    uses_stack.pop()
+                uses_stack[-1] = variable_list
+                if not isinstance(
+                    uses_stack[-2], (nodes.StmtWhileNode, nodes.StmtIfNode)
+                ):
+                    while True:
```

Comment:
nie uzywajcie prosze loopow wewnatrz DFSa

Reply:
te pętlę służą tylko do zdejmowania elementów ze stosu więc złożoność tworzenia słownika nadal jest liniowa

Reply:
Rozpisz mi wiec jaka dokladnie to ma zlozonosc. Nie wierze, ze while w while w DFSie to liniowa zlozonosc.

Reply:
Tu nie ma while w while tylko while są obok siebie i złożoność tego to ~2n, ponieważ podczas działania DFS dodawanych jest na stos n nodów, a w każdej iteracji while zdejmowany jest node ze stosu więc podczas całego działania DFSa niemożliwe jest aby petla while ziterowała się w sumie więcej niż n razy.
W dużym uproszczeniu działa to mniej więcej tak:
```
j = 0 
for i in range(0, n):
   ...
   if i == n-1:
      while j<n:
         j++
```
i w tym przypadku złożoność wynosi 2n tylko że różnica jest taka że pętla while nie wykonuje się cała dla ostatniej iteracji DFSa, tylko wykonuje się po trochu w konkretnych iteracjach DFS

Reply:
Używamy while po to aby zdejmować nody ze stosu, a zdejmujemy nody ze stosu po to aby wyłapać wszytskie zagnieżdżenia każdej zmiennej.

Reply:
While jest niepotrzebny w takim razie.

Reply:
to jak to zrobić?

Reply:
po prostu zdjac ze stacka element, jak wychodzisz z aktualnego elementu.

```
def dfs(node):
  # tutaj wlasnie wszedles do node
  print('> ' + node)
  
  for child in node.children:
    dfs(child)

  # tutaj wlasnie wyszedles z node
  print('< ' + node)

```